File: stratego/stratego-0.4.18/spec/bak.slib/spec/string.r
\literate[{\btt STRING}]

% Copyright (C) 1998, 1999, 2000 Eelco Visser <visser@acm.org>
% 
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2, or (at your option)
% any later version.
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
% 02111-1307, USA.

	This module defines some operations on strings, including
	conversions to and from numbers.

\begin{code}
module string
imports list conditional iteration

rules

  sect(op, arg) : x -> <op> (x, <arg>())

strategies
  
  new = prim("_ST_new")

  is-string = prim("_ST_is_string")

  implode-string = prim("_ST_implode_string")
  explode-string = prim("_ST_explode_string")

  conc-strings = (explode-string, explode-string); conc; implode-string

  concat-strings = map(explode-string); concat; implode-string

  int-to-string = 
    rec x(split(sect(mod, !10); sect(add, !48), sect(div, !10); int);
          (id, ?0; ![] <+ x); MkCons );
    reverse;
    implode-string

  string-to-int =
    explode-string;
    split(!0, id);
    repeat(S2I2); 
    S2I1

  escape = 
    explode-string;
    rec x(Escape; [id, id | x] <+ [id | x] <+ []);
    implode-string

rules

  Escape : [34 | cs] -> [92, 34 | cs]
  Escape : [92 | cs] -> [92, 92 | cs]

  S2I1 : (n, [])         -> n

  S2I2 : (n, [m|ms]) -> (<add>(<mul>(10, n), <subt>(m, 48)), ms)
         where <geq>(m, 48); <leq>(m, 57)

  S2D0 : (n, [46|ys])  -> (n, 10, ys)

  S2D1 : (n, f, []) -> n

  S2D2 : (n, f, [m|ms]) -> 
	 (<add>(n, <div>(<subt>(m, 48), f)), <mul>(f, 10), ms)
         where <geq>(m, 48); <leq>(m, 57)

strategies

  string-to-num =
    explode-string;
    split(!0, id);
    repeat(S2I2);
    ( S2I1
    + S2D0;
      repeat(S2D2);
      S2D1
    )

strategies

  lower-case = 
    explode-string;
    map(lc);
    implode-string

   lc = try(where(sect(geq, !65)); where(sect(leq, !90)); sect(add, !32))

rules

  SplitInit : x -> ([], [], x)

  SplitExit : 
	(xs, cs, []) -> 
	<reverse> [<reverse; implode-string> cs|xs]

  SplitNext : 
	(xs, cs, [32|ys]) -> 
	(Cons(<reverse; implode-string> cs, xs), [], ys)

  SplitNext : 
	(xs, cs, [y|ys]) -> (xs, [y|cs], ys)
	where <not(eq)> (y, 32)

strategies

  split-at-space =
    explode-string;
    SplitInit;
    rec x(SplitExit <+ SplitNext; x)  

  basename =
    explode-string;
    try(rec x([id|x] <+ ?[46 | _]; ![]));
    implode-string

  basename(ext) =
    explode-string;
    try(rec x([id|x] <+ [46 | ext]; ![]));
    implode-string

  guarantee-extension(ext) =
	basename;
	split(id, <ext>());
	add-extension

rules

  add-extension : (name, ext) -> <concat-strings> [name, ".", ext]
\end{code}

File: stratego/stratego-0.4.18/spec/bak.slib/src/string.c
/*

Copyright (C) 1998, 1999 Eelco Visser <visser@acm.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA.

*/

/* $Id: string.c,v 1.2 2000/01/19 16:24:15 visser Exp $ */

#include "stratego.h"
#include <stdlib.h>

/* Strings */

ATbool _is_string(void)
{ char *k;
  if(ATmatch(Ttop, "<str>", &k))
    {
      return ATtrue; 
    } 
  else return ATfalse; 
}

/* New: generating new strings */

int new_counter = 0;
int new_alphacounter = 0;
char new_string[256] = "";
#define ASCIIa 97

void new(void)
{
  sprintf(new_string, "%c_%d", ASCIIa + new_alphacounter, new_counter);
  while(AT_findSymbol(new_string, 0, ATtrue)) {
    new_alphacounter++;
    if(new_alphacounter % 26 == 0) {
      new_alphacounter = 0; 
      new_counter++;
    }
    sprintf(new_string, "%c_%d", ASCIIa + new_alphacounter, new_counter);
  }
  Tset((ATerm) ATmakeAppl0(ATmakeSymbol(new_string, 0, ATtrue)));
}

ATbool _implode_string(void)
{ 
  ATerm chars; 
  int i;
  
  /* Initial buffer size */
  static int size = 1024;
  static char* str = NULL;

  ATerm t;

  /* Allocate initial buffer */
  if( str == NULL )
  {
     str = (char*)malloc( size ); 
     assert( str != NULL );
  }

  chars = Ttop;
  /* ATfprintf(stderr, "chars = %t\n", chars); */
  for(i = 0; !AThasName(chars, "Nil"); 
      chars = (ATerm) ATgetNext((ATermList) chars), i++)
    {
    
     /* more room needed to store imploded string */
     if( i >= size )
     {
        /* double the allocated buffer */
        size *= 2;
        str = (char*)realloc( str, size );
        assert( str != NULL );
     }
     
      t = ATgetFirst((ATermList) chars);
      /* ATfprintf(stderr, "t = %t\n", t); */
      if(!ATisInt(t))
	return ATfalse;
      else
	str[i] = ATgetInt((ATermInt) t);
    }
  str[i] = '\0';

  /* Try to keep buffer size close to 1024 bytes whenever possible */
  if( size > 1024 )
  {
     size = i + 1;
     str = (char*)realloc( str, size);
  }
  Tset((ATerm) ATmakeString(str));
  return ATtrue;
}

ATbool _explode_string(void)
{ 
  char *str;
  int i;
  ATerm chars;
  if(!ATisString(Ttop))
    {
      /* ATfprintf(stderr, "warning: " 
	 "explode_string: term not a string: %t\n", Ttop); */
      /* assert(ATisString(Ttop)); */
      return ATfalse;
    }
  str = ATgetName(ATgetSymbol(Ttop));
  for(i = 0; str[i] != '\0'; i++) ;
  for(chars = ATmake("Nil"); i > 0; i--) {
    chars = App2("Cons", (ATerm) ATmakeInt(str[i - 1]), chars);
    // if(str[i - 1] == 92) i--;
  }
  Tset(chars);
  return ATtrue;
}
